bigtable/metrics.go

```diff
@@ -0,0 +1,479 @@
+	}
+
+	// Create default meter provider
+	internalMetricsConfig := &metricsConfigInternal{
+		project: project,
+		clientAttributes: []attribute.KeyValue{
+			attribute.String(monitoredResLabelKeyProject, project),
+			attribute.String(monitoredResLabelKeyInstance, instance),
+			attribute.String(metricLabelKeyClientUID, clientUID),
+			attribute.String(metricLabelKeyClientName, clientName),
+		},
+		instruments:    []builtInMetricInstruments{},
+		builtInEnabled: builtInEnabledDefault,
+	}
+
+	defaultExporter, err := newMonitoringExporter(ctx, internalMetricsConfig.project, opts...)
+	if err != nil {
+		return nil, err
+	}
+	internalMetricsConfig.exporter = defaultExporter
+
+	// Create default meter provider
+	defaultMp := otelmetric.NewMeterProvider(
+		otelmetric.WithReader(
+			otelmetric.NewPeriodicReader(
+				defaultExporter,
+				otelmetric.WithInterval(defaultSamplePeriod),
+			),
+		),
+	)
+
+	userProvidedMeterProviders := []*otelmetric.MeterProvider{}
+	if addr := os.Getenv("BIGTABLE_EMULATOR_HOST"); addr != "" {
+		// Do not emit metrics when emulator is being used
+		internalMetricsConfig.builtInEnabled = false
+	} else if userProvidedConfig != nil {
+		internalMetricsConfig.builtInEnabled = userProvidedConfig.builtInEnabled
+		userProvidedMeterProviders = userProvidedConfig.meterProviders
+	}
+
+	if !internalMetricsConfig.builtInEnabled {
+		return internalMetricsConfig, nil
+	}
+
+	// Create instruments on all meter providers
+	allMeterProviders := append(userProvidedMeterProviders, defaultMp)
+	for _, mp := range allMeterProviders {
+		builtInMetricInstruments := builtInMetricInstruments{
+			distributions: make(map[string]metric.Float64Histogram),
+			counters:      make(map[string]metric.Int64Counter),
+		}
+
+		// Create meter
+		meter := mp.Meter(instrumentationScope, metric.WithInstrumentationVersion(internal.Version))
+
+		// Create instruments
+		for metricName, metricDetails := range builtinMetrics {
+			if metricDetails.metricType == otelmetric.InstrumentKindHistogram {
+				builtInMetricInstruments.distributions[metricName], err = meter.Float64Histogram(
+					metricName,
+					metric.WithDescription(metricDetails.desc),
+					metric.WithUnit(metricDetails.unit),
+				)
+				if err != nil {
+					return internalMetricsConfig, err
+				}
+			} else if metricDetails.metricType == otelmetric.InstrumentKindCounter {
+				builtInMetricInstruments.counters[metricName], err = meter.Int64Counter(
+					metricName,
+					metric.WithDescription(metricDetails.desc),
+					metric.WithUnit(metricDetails.unit),
+				)
+				if err != nil {
+					return internalMetricsConfig, err
+				}
+			}
+			internalMetricsConfig.instruments = append(internalMetricsConfig.instruments, builtInMetricInstruments)
+		}
+	}
+	return internalMetricsConfig, nil
+}
+
+func (config *metricsConfigInternal) getOperationRecorder(ctx context.Context, table string, appProfile string, method string, isStreaming bool) (*builtinMetricsTracer, func()) {
```

Comment:
Its very unclear what the 2nd member of the return tuple is used for.
Please add some docs

Reply:
Removed the 2nd member